// isomorphic-relay-router/lib/IsomorphicRelayRouterContext.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _isomorphicRelay = require('isomorphic-relay');

var _isomorphicRelay2 = _interopRequireDefault(_isomorphicRelay);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactRouterRelay = require('react-router-relay');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var IsomorphicRelayRouterContext = (function (_RelayRouterContext) {
    (0, _inherits3.default)(IsomorphicRelayRouterContext, _RelayRouterContext);

    function IsomorphicRelayRouterContext() {
        (0, _classCallCheck3.default)(this, IsomorphicRelayRouterContext);
        return (0, _possibleConstructorReturn3.default)(this, (0, _getPrototypeOf2.default)(IsomorphicRelayRouterContext).apply(this, arguments));
    }

    (0, _createClass3.default)(IsomorphicRelayRouterContext, [{
        key: 'render',
        value: function render() {
            return _react2.default.createElement(_isomorphicRelay2.default.RootContainer, (0, _extends3.default)({}, this.props, {
                Component: this._routeAggregator,
                renderFailure: this.renderFailure,
                renderFetched: this.renderFetched,
                renderLoading: this.renderLoading,
                route: this._routeAggregator.route
            }));
        }
        // Static members are not inherited on <IE11. So, we have to redefine them.

    }]);
    return IsomorphicRelayRouterContext;
})(_reactRouterRelay.RelayRouterContext);

IsomorphicRelayRouterContext.displayName = 'IsomorphicRelayRouterContext';
IsomorphicRelayRouterContext.propTypes = _reactRouterRelay.RelayRouterContext.propTypes;
IsomorphicRelayRouterContext.childContextTypes = _reactRouterRelay.RelayRouterContext.childContextTypes;
IsomorphicRelayRouterContext.defaultProps = _reactRouterRelay.RelayRouterContext.defaultProps;
exports.default = IsomorphicRelayRouterContext;
